### DISCUSSION TITLE
Digest without timestamp

### DISCUSSION THREAD

**Discussion by softpuff** [oras-project/oras-go#703] [Q&A]
Is it possible to have digest for pushed image that doesn't include timestamp so it's actually the same if the layers being pushed are the same?

**Comment by LarryHu0217** [comment]
If by “digest” you mean the pushed image/manifest digest, it cannot be determined from the layers alone. The digest is calculated from the manifest bytes; the manifest also references a config blob and may contain annotations. Identical layer digests therefore do not guarantee an identical manifest digest.

With oras-go `PackManifest`, a missing `org.opencontainers.image.created` annotation is filled with the current UTC time. For reproducible output, provide a fixed RFC 3339 timestamp:

```go
opts := oras.PackManifestOptions{
    Layers: layers,
    ManifestAnnotations: map[string]string{
        ocispec.AnnotationCreated: "1970-01-01T00:00:00Z",
    },
}
```

Keep the config descriptor/bytes and all other manifest inputs identical as well. A fixed timestamp makes that field deterministic; it does not make different manifest content produce the same digest. If you need a digest of only the layers, calculate and store that separately from the OCI image digest.

Sources: [oras-go `PackManifest` documentation and implementation](https://github.com/oras-project/oras-go/blob/v2.6.0/pack.go#L90-L140), [OCI image manifest specification](https://github.com/opencontainers/image-spec/blob/v1.1.1/manifest.md).
